### ISSUE TITLE
Can I replace CUDA_VM features to previous CUDA calls?

### ISSUE THREAD

**Issue by Esthesia** [NVIDIA/DALI#4779]
In `DALI/dali/core/mm/default_resources.cc` file, there is a code on using cuda_vm features.

```
...
  #if DALI_USE_CUDA_VM_MAP
  if (cuvm::IsSupported() && UseVMM()) {
    using resource_type = mm::async_pool_resource<mm::memory_kind::device, cuda_vm_resource,
                                                  std::mutex, void>;
    return std::make_shared<resource_type>();
  }
  #endif  // DALI_USE_CUDA_VM_MAP
...
```

I know CUDA_VM features are good at dealing GPU memory, but could I bypass this VM feature and use `cuda_malloc_memory_resource`?

**Comment by JanuszL** [comment]
Hi @Esthesia,

You can try but as far as I know, all platforms DALI supports now use CUDA_VM, and we are not testing this code path. Still, you can try.
Also, could you tell me what you plan to achieve? This code path means to speed up the GPU memory allocation process. Other code paths could be very inefficient.

**Comment by Esthesia** [comment]
Hi @JanuszL, thanks for quickly replying.

I try to hook the CUDA malloc call so I consider not using the CUDA_VM feature (which is based on reserve and mapping functionality on CUDA) for my simplicity.

**Comment by mzient** [comment]
Hello @Esthesia 
Just set the environment variable `DALI_USE_VMM=0` and you're good. That said, this is _not_ recommended because cudaMalloc introduces a lot of unnecessary synchronization whenever we need to increase the pool size _and_ it can easily trigger out-of-memory due to fragmentation. *The memory consumption is is _much higher_ when not using VMM* - depending on your workload, this may be a few percent or it may be 3x. Large (and highly variable) batches exacerbate the problem. Depending on what's the goal of this hooking, not using VMM may just defeat the purpose.

**Comment by Esthesia** [comment]
Hello @mzient, thanks for the detailed illustration.
It is great to know that highly variable increases in memory consumption.
Is the main reason for fragmentation due to the coarse-grained memory allocating on cudaMalloc
that naively allocates the requested amount of memory?

**Comment by Esthesia** [comment]
Hello @JanuszL,

> In the case of cudaMalloc subsequent allocations may not necessary be adjacent

Oh, I thought that one request got consequent allocation but it seems it is not necessary.
Good to know that for me.

Thank you for detailed answer.

**Comment by JanuszL** [comment]
> Oh, I thought that one request got consequent allocation but it seems it is not necessary.

I didn't mean that. What I mean is that if you call cudaMalloc twice, and you get 0-10, and 11-20 addresses, then you cannot treat this as one allocation spanning 0-20, as some calls will detect that and error. In the case of VMM we allocate 0-100 (just an example) and then just fill it with the physical pages, so the access 0-20 can be valid, while 21-100 are not backed up and will error out.

**Comment by Esthesia** [comment]
Now I got the point. Thanks for the correction

**Comment by mzient** [comment]
About fragmentation with cudaMalloc - consider the following scenario:
The program requests the following blocks of memory (in sequence):

0.8 GB
2.5 GB
0.5 GB
1 GB

The memory manager cannot predict that and it'll allocate a 1GB slab for the first allocation and then 4GB slab for the second allocation. The third and fourth allocations can be accommodated by the second slab.
Now if the user frees the 2nd (2.5GB) and 4th (1GB) blocks the 4GB slab ends up with an unmovable 0.5GB block in the middle. There's 3.5GB of free memory in the slab, but the largest available block is just 2.5GB. Now if the user requests a 3GB block, the memory manager will have to allocate _another_ slab. The memory consumption is very high, but the actual demand for memory is much lower - it's the fragmentation that drives the demand.

With CUDA VMM, we'd allocate some virtual address space. The allocations would request several fixed-sized physical blocks and map them. There are no "slabs".
When the blocks are freed, the subsequent allocations can either try to reuse the free areas directly (fastest route) or, if there's no contiguous block that's large enough, the memory manager will unmap some unused physical blocks and reassign them to the new region (and possibly allocate more physical blocks, if required). With this approach, the fragmentation is greatly reduced. Of course, it's possible to construct a pessimistic scenario (e.g. allocate 1000 small blocks and then keep every tenth block, so that each physical block is at least partially used) - but it's unlikely to occur in a realistic setting and it's partially mitigated by overaligning large allocations so that they start at a physical block boundary.
